src/main.rs

```diff
@@ -58,21 +59,55 @@ impl LanguageServer for Backend {
             .to_file_path()
             .unwrap()
             .with_extension("pdf");
-
+        let messages: Vec<_>;
         let output = match typst::compile(world) {
             Ok(document) => {
                 let buffer = typst::export::pdf(&document);
+                messages = vec![];
                 fs::write(output_path, buffer).map_err(|_| "failed to write PDF file".to_string())
             }
             Err(errors) => {
-                let messages: Vec<_> = errors.iter().map(|error| error.message.as_str()).collect();
-                Err(messages.join("\n"))
+                messages = errors
+                    .iter()
+                    .map(|error| (world.source(error.span.source()), error))
+                    .map(|(source, se)| {
+                        let range = source.range(se.span);
+                        (
+                            se.message.to_string(),
+                            Range {
+                                start: Position {
+                                    line: source.byte_to_line(range.start).unwrap() as _,
+                                    character: source.byte_to_column(range.start).unwrap() as _,
+                                },
+                                end: Position {
+                                    line: source.byte_to_line(range.end).unwrap() as _,
+                                    character: source.byte_to_column(range.end).unwrap() as _,
+                                },
+                            },
+                        )
+                    })
```

Comment:
Extract this into a function so we just need to call `.iter().map(the_function).collect()` or `.iter().map(|error| something.the_function(error)).collect()`.

It will be easier to read `did_change` and understand this chunk of code.